### PR TITLE
CXP-1390: Constraint - string - min / max length

### PR DESCRIPTION
**Pull request by guillaumefer-akeneo**
**Description (for Contributor and Core Developer)**
image
**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

**Review by eoliveauakeneo — APPROVED** [review]

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+          "$ref": "#/$defs/description"
+        },
+        "type": {
+          "enum": [
+            "string",
+            "number",
+            "boolean"
+          ]
+        },
+        "minLength": {
```
Comment (RiffVentura) [inline components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json:82]: How permissive should we be in the accepted schema? 
With this implementation this is a valid property:
```JSON
"body_html": {
  "title": "Description",
  "description": "Une description",
  "type": "boolean", //<====== boolean with min length
  "minLength": 2,
  "maxLength": 3
}
```

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+          "enum": [
+            "string"
+          ]
+        },
+        "format": {
+          "enum": [
+            "date-time"
+          ]
+        },
+        "minLength": {
```
Comment (RiffVentura) [inline components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json:115]: IMO it should be refactored so that we won't have to define it more than once

File: components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php [hunk 3]
```diff
@@ -190,6 +197,11 @@ private function getProductMappingSchemaRaw(): string
             "meta_title": {
               "type": "string",
               "title": "Meta title"
+            },
+            "meta_description": {
+              "type": "string",
+              "minLength": 50,
```
Comment (RiffVentura) [inline components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php:203]: :thinking:  I'm wondering if all requirements should be under a sub-property...

Comment (tseho) [reply components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php:203]: what do you mean ?

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json (hunk 2 above)
Comment (masacc) [reply components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json:115]: Another solution was proposed by @hogren and I implemented it in my next version of the meta schema. We have to synchronize between half squads about meta schema updates.

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json (hunk 1 above)
Comment (tseho) [reply components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json:82]: Yes, it look dumb but the official spec allows it, they are simply ignored when added on the wrong type.

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json (hunk 2 above)
Comment (masacc) [reply components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json:115]: here is the proposal : https://github.com/akeneo/pim-api-docs/pull/749

File: components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php (hunk 3 above)
Comment (RiffVentura) [reply components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php:203]:
```JSON
"body_html": {
  "title": "Description",
  "description": "Une description",
  "type": "boolean",
  "constraints": {
    "minLength": 2,
    "maxLength": 3
  },
}
```

the thought process here is that if there are a lot of requirements, it might be hard to read/grab only constraints properties

Comment (tseho) [reply components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php:203]: They did it like in the spec https://json-schema.org/understanding-json-schema/reference/string.html#length

Comment (tseho) [reply components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php:203]: I'm not sure we have something to gain if we deviate from the official spec.
Power users that knows how to write a json schema will have to learn again.

File: components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php [hunk 4]
```diff
@@ -120,6 +120,11 @@ protected function execute(InputInterface $input, OutputInterface $output): int
                     'scope' => null,
                     'locale' => 'en_US',
                 ],
+                'meta_description' => [
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php:123]: Why adding this new property in the fixtures ? for me it's the same as "name" in terms of use case (string with min/max length)

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        }
+      },
+      "additionalProperties": false,
+      "required": [
+        "type"
+      ]
+    }
+  },
+  "additionalProperties": false
+}
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json:165]: Here is the version with the `dependentSchema` solution proposed by @hogren :
```suggestion
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://api.akeneo.com/mapping/product/0.0.5/schema",
  "type": "object",
  "properties": {
    "$id": { "$ref": "#/$defs/$id" },
    "$schema": { "$ref": "#/$defs/$schema" },
    "$comment": { "$ref": "#/$defs/$comment" },
    "$defs": { "$ref": "#/$defs/$defs" },
    "title": { "$ref": "#/$defs/title" },
    "description": { "$ref": "#/$defs/description" },
    "type": { "const": "object" },
    "properties": {
      "type": "object",
      "properties": {
        "uuid": { "$ref": "#/$defs/uuidProperty" }
      },
      "additionalProperties": { "$ref": "#/$defs/property" },
      "default": {},
      "required": ["uuid"]
    }
  },
  "$defs": {
    "$id": {
      "$comment": "Non-empty fragments not allowed.",
      "pattern": "^[^#]*#?$"
    },
    "$schema": { "$ref": "#/$defs/uriString" },
    "$comment": {
      "type": "string"
    },
    "$defs": {
      "type": "object",
      "additionalProperties": { "$ref": "#/$defs/property" }
    },
    "title": {
      "type": "string"
    },
    "description": {
      "type": "string"
    },
    "property": {
      "type": "object",
      "properties": {
        "title": { "$ref": "#/$defs/title" },
        "description": { "$ref": "#/$defs/description" },
        "type": {
          "enum": [
            "string",
            "number",
            "boolean"
          ]
        },
        "format": {
          "enum": [
            "date-time"
          ]
        }
      },
      "additionalProperties": false,
      "required": ["type"],
      "dependentSchemas": {
        "format": {
          "properties": {
            "type": {
              "enum": [
                "string"
              ]
            }
          },
          "required": ["type"]
        }
      },
      "minLength": {
          "type": "integer",
          "minimum": 0,
          "default": 0
        },
        "maxLength": {
          "type": "integer",
          "minimum": 0
        }
    },
    "uriString": {
      "type": "string",
      "format": "uri"
    },
    "uuidProperty": {
      "type": "object",
      "properties": {
        "title": { "$ref": "#/$defs/title" },
        "description": { "$ref": "#/$defs/description" },
        "type": { "const": "string" }
      },
      "additionalProperties": false,
      "required": ["type"]
    }
  },
  "additionalProperties": false
}
```

File: components/catalogs/back/tests/Integration/Infrastructure/Validation/ProductSchema/invalid/0.0.5-with-format-for-wrong-type.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
+{
+  "$schema": "https://api.akeneo.com/mapping/product/0.0.5/schema",
+  "description": "The data should match one item from enum at #/properties/name/type",
+  "properties": {
+    "uuid": {
+      "type": "string"
+    },
+    "name": {
+      "type": "array"
+    },
```
Comment (masacc) [inline components/catalogs/back/tests/Integration/Infrastructure/Validation/ProductSchema/invalid/0.0.5-with-format-for-wrong-type.json:10]: I fixed it for version 0.0.4 in another PR, you should do it here too :
- rm this line 
- change the description to "The data should match one item from enum at #/properties/release_date/type" (with the new schema I proposed)

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    },
+    "title": {
+      "$ref": "#/$defs/title"
+    },
+    "description": {
+      "$ref": "#/$defs/description"
+    },
+    "type": {
+      "const": "object"
+    },
```
Comment (hogren) [inline components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json:26]: Why did you change the style of this block ? I am not agree nor disagree. I just think it disturb the diff analysis and we need to debate if you think that is better than the previous. @akeneo/squad-octopus What do you think, do we need to keep the new syntax or the old which was :

>     "$id": { "$ref": "#/$defs/$id" },
>     "$schema": { "$ref": "#/$defs/$schema" },
>     "$comment": { "$ref": "#/$defs/$comment" },
>     "$defs": { "$ref": "#/$defs/$defs" },
>     "title": { "$ref": "#/$defs/title" },
>     "description": { "$ref": "#/$defs/description" },
>     "type": { "const": "object" },

File: components/catalogs/front/src/components/ProductMapping/components/RequirementsCollapse.tsx [hunk 8]
```diff
@@ -2,16 +2,41 @@ import React, {FC, useState} from 'react';
...
+            constraints.push({
+                key: targetKey,
+                value: target[targetKey],
+            } as Constraint);
+        }
+    });
+
+    const shouldDisplayWarning = constraints.length > 0;
+
+    if ((undefined === target.description || null === target.description) && !shouldDisplayWarning) {
```
Comment (hogren) [inline components/catalogs/front/src/components/ProductMapping/components/RequirementsCollapse.tsx:39]:
```suggestion
    if (!shouldDisplayWarning) {
```
Do you have a contraindication for this ?

**Comment by hogren** [comment]
I saw a strange graphic effect.
Sometimes, the requirements block appears first like that :
image

And after 1 or 2 seconds, a little space appears at the bottom :
image

**Comment by hogren** [comment]
The alignment of the icons is not very good, if I look at the design mockup.
In mockup :
image

In real :
image

Can you try to adjust ?

**Comment by helenegzlz** [comment]
Hi @hogren, about the icon's alignment the issue has been seen with Sophie in the CXP-1387, it stays as it is in the DSM (so not vertically centered) :) 
image

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json (hunk 5 above)
Comment (helenegzlz) [reply components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json:165]: @masacc did it in another PR.

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json (hunk 7 above)
Comment (masacc) [reply components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json]: Keep the old ones, this part is not the most important and is readable as it was

File: components/catalogs/back/tests/Integration/Infrastructure/Validation/ProductSchema/invalid/0.0.4-with-format-for-wrong-type.json [hunk 9]
```diff
@@ -1,13 +1,10 @@
 {
-  "$schema": "https://api.akeneo.com/mapping/product/0.0.4/schema",
-  "description": "The data should match one item from enum at #/properties/name/type",
+  "$schema": "https://api.akeneo.com/mapping/product/0.0.5/schema",
```
Comment (masacc) [inline components/catalogs/back/tests/Integration/Infrastructure/Validation/ProductSchema/invalid/0.0.4-with-format-for-wrong-type.json:2]:
```suggestion
  "$schema": "https://api.akeneo.com/mapping/product/0.0.4/schema",
```

File: components/catalogs/front/src/components/ProductMapping/models/Target.d.ts [hunk 10]
```diff
@@ -1,7 +1,10 @@
 export type Target = {
+    [key: string]: any;
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductMapping/models/Target.d.ts:2]: why ?

File: components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php (hunk 4 above)
Comment (helenegzlz) [reply components/catalogs/back/src/Infrastructure/Command/CatalogFixtureCommand.php]: The property had been removed in this commit : f002e5dc2c1c7bdb2d19ef1d2c638ffbfd3d79e4

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json (hunk 7 above)
Comment (helenegzlz) [reply components/catalogs/back/src/Infrastructure/Symfony/Resources/meta-schemas/product-0.0.5.json]: Fixed in the f002e5dc2c1c7bdb2d19ef1d2c638ffbfd3d79e4 commit

File: components/catalogs/back/tests/Integration/Infrastructure/Validation/ProductSchema/invalid/0.0.4-with-format-for-wrong-type.json (hunk 9 above)
Comment (helenegzlz) [reply components/catalogs/back/tests/Integration/Infrastructure/Validation/ProductSchema/invalid/0.0.4-with-format-for-wrong-type.json]: Fixed here 8309b025fd42c4e6593bbffd6f9124b84c5061b7

File: components/catalogs/front/src/components/ProductMapping/models/Target.d.ts (hunk 10 above)
Comment (hogren) [reply components/catalogs/front/src/components/ProductMapping/models/Target.d.ts:2]: @masacc  It's a way to tell to TS to accept the accessors of type []. e.g `selectedTarget[acceptedConstraint]` . I need it to make a clear code at lines 28-39.

Comment (tseho) [reply components/catalogs/front/src/components/ProductMapping/models/Target.d.ts:2]: Yeah, that's not good, it basically tells typescript you can put anything you want in this object.
give me 10min to think about an alternative.

Comment (tseho) [reply components/catalogs/front/src/components/ProductMapping/models/Target.d.ts:2]: I got interrupted.
Anyway, here is a proposal. https://github.com/akeneo/pim-community-dev/pull/19141

**Review by masacc — APPROVED** [review]
